### PR TITLE
Issue#461

### PR DESCRIPTION
**Pull request by stepanzharychevbroadcom**
Implementation for use cases listed in https://github.com/zowe/vscode-extension-for-zowe/issues/461#issuecomment-578194128

### CODE REVIEW

File: src/utils/uss.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * Injects extra data to tooltip based on node status and other conditions
+ * @param node
+ * @param tooltip
+ * @returns {string}
+ */
+export function injectAdditionalDataToTooltip(node: ZoweUSSNode, tooltip: string) {
+    if (node.downloaded && node.downloadedTime) {
+        // TODO: Add time formatter to localization so we will use not just US variant
+        return `${tooltip} (Last Pulled: ${moment(node.downloadedTime).format("HH:mm MM/DD/YY")})`;
```
Comment (Colin-Stone) [inline src/utils/uss.ts]: Really useful tooltip. Just a question. "Last pulled" or "Last downloaded"

Comment (stepanzharychevbroadcom) [reply src/utils/uss.ts]: I guess they have equal meaning. First is just shorter.

Comment (Colin-Stone) [reply src/utils/uss.ts]: I think Downloaded would probably mean more to most users. It could always be simply "Downloaded" instead of "Last Pulled" ?.

Comment (stepanzharychevbroadcom) [reply src/utils/uss.ts]: Yes, sure we can.

**Comment by stepanzharychevbroadcom** [comment]
> Really neat function.I did find something weird though. I downloaded a UTF-8 file. Which was converted into gibberish. Hit the Toggle Binary button and everything was perfect, icon was right etc. When I clicked the file again it redownloaded it in gibberish. i.e. didn't maintain the binary decision. Same thing if I use the cloud download button.
> 
> Got a `Cannot read property 'setEtag' of undefined` popup when saving a file I had switched to binary

@Colin-Stone I tried it with image and it works. In the case if you changed your filter (e.g. from one folder to another) this decision won't be kept, it's not yet implemented.
For "Downloaded text" I added fix.

**Comment by stepanzharychevbroadcom** [comment]
@Colin-Stone @Alexandru-Dumitru Could you take a look again? I would really like to have it merged before more changes in master will happen.

**Comment by Colin-Stone** [comment]
Is it me or have the icon defintion changed?. They are paler, less well defined than those around them and resemble the reversed icons technique I use for indicating migrated datasets currently. i.e the normal file icon for a dark theme used in the light theme and vice-versa. These are the file (text and binary ones)
image 
image
image

**Comment by stepanzharychevbroadcom** [comment]
> Is it me or have the icon defintion changed?. They are paler, less well defined than those around them and resemble the reversed icons technique I use for indicating migrated datasets currently. i.e the normal file icon for a dark theme used in the light theme and vice-versa. These are the file (text and binary ones)
> image
> image
> image

@Colin-Stone No, that's how it should be, once the file is downloaded we decrease the opacity of icon so it will be additionally understandable for the user that the file is different. It's 30% opacity right now. We came to this solution together with UI designer.

File: src/ZoweUSSNode.ts [hunk 2]
```diff
@@ -216,22 +232,69 @@ export class ZoweUSSNode extends vscode.TreeItem implements IZoweTreeNode {
...
     }
+
+    /**
+     * Helper getter to check dirtiness of node inside opened editor tabs, can be more accurate than saved value
+     *
+     * @returns {boolean}
+     */
+    public get isDirtyInEditor(): boolean {
+        const openedTextDocuments = vscode.workspace.textDocuments;
+        const currentFilePath = getUSSDocumentFilePath(this);
```
Comment (Alexandru-Dumitru) [inline src/ZoweUSSNode.ts]: Why not `extension.getUSSDocumentFilePath(this);` and avoid duplicate imports?

File: src/ZoweUSSNode.ts [hunk 3]
```diff
@@ -216,22 +232,69 @@ export class ZoweUSSNode extends vscode.TreeItem implements IZoweTreeNode {
...
+                return document.isDirty;
+            }
+        }
+
+        return false;
+    }
+
+    public get openedDocumentInstance(): vscode.TextDocument {
+        const openedTextDocuments = vscode.workspace.textDocuments;
+        const currentFilePath = getUSSDocumentFilePath(this);
```
Comment (Alexandru-Dumitru) [inline src/ZoweUSSNode.ts]: same here

File: src/ZoweUSSNode.ts [hunk 4]
```diff
@@ -12,11 +12,21 @@
 import * as zowe from "@brightside/core";
 import { Session } from "@brightside/imperative";
 import * as vscode from "vscode";
+// tslint:disable-next-line: no-duplicate-imports
+import { TreeItemCollapsibleState } from "vscode";
```
Comment (Alexandru-Dumitru) [inline src/ZoweUSSNode.ts]: This line is redundant, because `TreeItemCollapsibleState` is never used directly.

File: src/ZoweUSSNode.ts [hunk 5]
```diff
@@ -12,11 +12,21 @@
...
 import { IZoweTreeNode } from "./ZoweTree";
-const localize = nls.config({ messageFormat: nls.MessageFormat.file })();
+// tslint:disable-next-line: no-duplicate-imports
 import * as extension from "../src/extension";
+// tslint:disable-next-line: no-duplicate-imports
+import { getUSSDocumentFilePath } from "../src/extension";
 import * as utils from "./utils";
+import { getIconByNode } from "./generators/icons/index";
+// tslint:disable-next-line: no-implicit-dependencies
+import * as moment from "moment";
```
Comment (Alexandru-Dumitru) [inline src/ZoweUSSNode.ts]: Shouldn't we add this as dependency instead of `momentjs` ?

File: src/ZoweUSSNode.ts [hunk 6]
```diff
@@ -250,10 +308,37 @@ export class ZoweUSSNode extends vscode.TreeItem implements IZoweTreeNode {
...
+         * helper method to change the node names in one go
+         * @param oldReference string
+         * @param revision string
+         */
+    }
+
     /**
-     * helper method to change the node names in one go
-     * @param oldReference string
-     * @param revision string
```
Comment (Alexandru-Dumitru) [inline src/ZoweUSSNode.ts]: Not sure why part of documentation jumped inside the body of the function 😄

Comment (stepanzharychevbroadcom) [reply src/ZoweUSSNode.ts]: @Alexandru-Dumitru Fixed.

**Pull request merged**